### ISSUE TITLE
Option for squash and merge to also make a merge commit

### ISSUE THREAD

**Issue by JacobEvelyn** [isaacs/github#1328]
We like our commit history to be clean. During code review it's not uncommon for the first commit to have a useful commit message while other commits in the PR have messages like `update code for code review comments`. Before merging, to keep our history clean, we squash those extra commits into the first one.

The annoying thing is that since we want GitHub to enforce that PRs pass tests on CI before we merge them, we have to wait for tests to pass _again_ after squashing, even though the code has not changed at all. (Note that this is different from [rebasing, where the code _has_ changed](https://github.com/isaacs/github/issues/1143#issuecomment-402892640).) Since at this point we know that tests will pass, it would be nice if we could just squash and merge at the click of a button and not have to wait for tests to run _again_.

Fortunately, we can, with GitHub's "squash and merge" option!

![34022126-5662c81c-e0f2-11e7-9237-3446cb429ae4](https://user-images.githubusercontent.com/1114569/44234880-b71a4d80-a175-11e8-9941-b37dc4899f59.png)

Unfortunately, our deployment tooling that specifically looks for merge commits, and it would be difficult for us to rewrite all of it. So we're stuck squashing on our local machines, force-pushing to our branches, and then waiting for tests to re-run before we merge PRs. But if GitHub allowed "squash and merge" to make a merge commit, our problems would go away and we could spend more time shipping features and less time doing manual squashes and waiting on CI.

**Comment by AraHaan** [comment]
I do not like this one bit, on one hand some people love merge commits, on another hand people hate it.

As such some employers even fire people just for making merge commits because they add nothing and just add noise and so they prefer rebase and squash merges instead to avoid them.

While this, and another issue related to doing this to rebase can be useful to some, I think it should be off by default and enabled by a repository setting instead named ``Enable merge commits from Rebase and Sqaush``. This would ensure no breakage from current repositories that expect the current behavior while also allowing people who want it to have it.

**Comment by aspiers** [comment]
[**@AraHaan**](https://github.com/AraHaan) commented on [April 28, 2021 3:36 AM](https://github.com/isaacs/github/issues/1328#issuecomment-828096638 "2021-04-28T02:36:09Z - Replied by Github Reply Comments"):
> As such some employers even fire people just for making merge commits because they add nothing and just add noise and so they prefer rebase and squash merges instead to avoid them.

:exploding_head:  If I was working for an employer like that, I would quit before getting fired:

  1. That's a *completely* unacceptable reason for firing anyone.
  2. An employer who thinks merge commits add nothing but noise is clearly unwilling or incapable of understanding git properly.

> I do not like this one bit, on one hand some people love merge commits, on another hand people hate it.

Exactly, there can be valid reasons for both approaches.  In general good tools should not assume they always know better than their users.
